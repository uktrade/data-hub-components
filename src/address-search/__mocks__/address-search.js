import MockAdapter from 'axios-mock-adapter'
import axios from 'axios'

import {
  addressSearch,
  badRequest,
} from '../__fixtures__'

export function setupSuccessMocks(apiEndpoint) {
  const mock = new MockAdapter(axios)
  mock
    .onGet(apiEndpoint)
    .reply(200, addressSearch)
}

export function setupBadRequest(apiEndpoint) {
  const mock = new MockAdapter(axios)
  mock
    .onGet(apiEndpoint)
    .reply(400, badRequest)
}

export function setupErrorMocks(apiEndpoint) {
  const mock = new MockAdapter(axios)
  mock
    .onGet(apiEndpoint)
    .reply(500)
}

export function mockAddressVariation1(apiEndpoint) {
  const mock = new MockAdapter(axios)
  mock
    .onGet(apiEndpoint)
    .reply(200, {
      Addresses: [
        'zero, one, two, , , five, six',
      ],
    })
}

export function mockAddressVariation2(apiEndpoint) {
  const mock = new MockAdapter(axios)
  mock
    .onGet(apiEndpoint)
    .reply(200, {
      Addresses: [
        'zero, one, two, , , , ',
      ],
    })
}

export function mockAddressVariation3(apiEndpoint) {
  const mock = new MockAdapter(axios)
  mock
    .onGet(apiEndpoint)
    .reply(200, {
      Addresses: [
        'zero, one, , , , five, ',
      ],
    })
}
