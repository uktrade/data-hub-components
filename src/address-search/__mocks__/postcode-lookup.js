import MockAdapter from 'axios-mock-adapter'
import axios from 'axios'

import addressSearch from '../__fixtures__/address-search-SW1H 9AJ.json'

export function setupPostcodeMock200(apiEndpoint, adapterOptions = {}) {
  const mock = new MockAdapter(axios, adapterOptions)
  mock.onGet(apiEndpoint).reply(200, addressSearch)

  return mock
}

export function setupPostcodeMock400(apiEndpoint, adapterOptions = {}) {
  const mock = new MockAdapter(axios, adapterOptions)
  mock.onGet(apiEndpoint).reply(400)

  return mock
}

export function setupPostcodeMock404(apiEndpoint, adapterOptions = {}) {
  const mock = new MockAdapter(axios, adapterOptions)
  mock.onGet(apiEndpoint).reply(404)

  return mock
}
