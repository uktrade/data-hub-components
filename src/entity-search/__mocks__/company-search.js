import MockAdapter from 'axios-mock-adapter'
import axios from 'axios'

import fixtures from '../__fixtures__'

const mock = new MockAdapter(axios)

export function setupSuccessMocks(apiEndpoint) {
  mock
    .onPost(apiEndpoint, {})
    .reply(200, fixtures.companySearch)
  mock
    .onPost(apiEndpoint, { search_term: 'some other company' })
    .reply(200, fixtures.companySearchFilteredByCompanyName)
  mock
    .onPost(apiEndpoint, { postal_code: 'BN1 4SE' })
    .reply(200, fixtures.companySearchFilteredByPostcode)
}

export function setupErrorMocks(apiEndpoint) {
  mock.onPost(apiEndpoint, {}).reply(500)
  mock.onPost(apiEndpoint, { search_term: 'some other company' }).reply(500)
  mock.onPost(apiEndpoint, { postal_code: 'BN1 4SE' }).reply(500)
}

export function setupNoResultsMocks(apiEndpoint) {
  mock
    .onPost(apiEndpoint, {})
    .reply(200, fixtures.companySearchNoResults)
  mock
    .onPost(apiEndpoint, { search_term: 'some other company' })
    .reply(200, fixtures.companySearchNoResults)
  mock
    .onPost(apiEndpoint, { postal_code: 'BN1 4SE' })
    .reply(200, fixtures.companySearchNoResults)
}
