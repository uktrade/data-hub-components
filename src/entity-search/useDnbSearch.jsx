/* eslint-disable camelcase */

import React from 'react'
import axios from 'axios/index'
import { compact } from 'lodash'
import Link from '@govuk-react/link'

function useDnbSearch(apiEndpoint) {
  function getText(dnb_company, datahub_company) {
    if (dnb_company.is_out_of_business) {
      return 'This company has stopped trading and is no longer in business.'
    }

    if (datahub_company) {
      return (
        <>
          This company is already on Data Hub.
          {' '}
          <Link href={`/companies/${datahub_company.id}`}>Go to the company page</Link> to record activity.
        </>
      )
    }

    return null
  }

  function transformCompanyRecord(record) {
    const { dnb_company, datahub_company } = record

    return {
      id: dnb_company.duns_number,
      heading: dnb_company.primary_name,
      meta: {
        Address: compact([
          dnb_company.address_line_1,
          dnb_company.address_line_2,
          dnb_company.address_town,
          dnb_company.address_county,
          dnb_company.address_postcode,
        ]).join(', '),
      },
      text: getText(dnb_company, datahub_company),
      canHandleClick: !datahub_company,
      data: record,
    }
  }

  async function findCompany(filters) {
    const { data } = await axios.post(apiEndpoint, filters)

    return data.results.map(transformCompanyRecord)
  }

  return {
    findCompany,
    transformCompanyRecord,
  }
}

export default useDnbSearch
