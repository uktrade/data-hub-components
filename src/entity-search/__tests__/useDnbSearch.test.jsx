import { renderHook } from '@testing-library/react-hooks'

import useDnbSearch from '../useDnbSearch'
import searchResultsFixture from '../__fixtures__/company-search'
import { setupErrorMocks, setupSuccessMocks } from '../__mocks__/company-search'

const API_ENDPOINT = 'http://localhost:8000/v4/dnb/company-search'

describe('useDnbSearch', () => {
  const { result } = renderHook(() => useDnbSearch(API_ENDPOINT))
  let axiosMock
  let response

  describe('when transformCompanyRecord() is called', () => {
    let transformedCompanyRecord

    beforeAll(() => {
      transformedCompanyRecord = result.current.transformCompanyRecord(
        searchResultsFixture.results[0],
      )
    })

    test('should return transformed company record', () => {
      expect(transformedCompanyRecord).toMatchSnapshot()
    })
  })

  describe('when findCompany() is called without any filters', () => {
    beforeAll(async () => {
      axiosMock = setupSuccessMocks(API_ENDPOINT)
      response = await result.current.findCompany()
    })

    test('should return all DNB search results', () => {
      expect(response).toMatchSnapshot()
    })
  })

  describe('when findCompany() is called with "search_term" filter', () => {
    beforeAll(async () => {
      axiosMock = setupSuccessMocks(API_ENDPOINT)
      response = await result.current.findCompany({
        search_term: 'some other company',
      })
    })

    test('should call endpoint with the specified filter', () => {
      expect(axiosMock.history.post[0].data).toEqual(JSON.stringify({
        search_term: 'some other company',
      }))
    })

    test('should return only "Some company name" company', () => {
      expect(response).toMatchSnapshot()
    })
  })

  describe('when findCompany() is called on endpoint which returns error', () => {
    beforeAll(async () => {
      axiosMock = setupErrorMocks(API_ENDPOINT)

      try {
        response = await result.current.findCompany()
      } catch (e) {
        response = e
      }
    })

    test('should return an error message', () => {
      expect(response.toString()).toEqual(
        'Error: Request failed with status code 500',
      )
    })
  })
})
