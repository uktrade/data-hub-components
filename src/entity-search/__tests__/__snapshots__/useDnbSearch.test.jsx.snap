// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDnbSearch when findCompany() is called with "search_term" filter should return only "Some company name" company 1`] = `
Array [
  Object {
    "canHandleClick": true,
    "data": Object {
      "datahub_company": null,
      "dnb_company": Object {
        "address_country": "GB",
        "address_county": "",
        "address_line_1": "123 ABC Road",
        "address_line_2": "",
        "address_postcode": "BN2 9QB",
        "address_town": "Brighton",
        "annual_sales": 1999999999,
        "annual_sales_currency": "USD",
        "domain": "example.co.uk",
        "duns_number": "219999999",
        "employee_number": 300,
        "global_ultimate_duns_number": "319999999",
        "global_ultimate_primary_name": "Some other company parent",
        "industry_codes": Array [
          Object {
            "usSicV4": "3799",
            "usSicV4Description": "Mfg transportation equipment",
          },
        ],
        "is_annual_sales_estimated": null,
        "is_employees_number_estimated": true,
        "is_out_of_business": false,
        "legal_status": "corporation",
        "primary_name": "Some other company",
        "registered_address_country": "GB",
        "registered_address_county": "",
        "registered_address_line_1": "",
        "registered_address_line_2": "",
        "registered_address_postcode": "BN2 9QB",
        "registered_address_town": "Brighton",
        "registration_numbers": Array [
          Object {
            "registration_number": "00016033",
            "registration_type": "uk_companies_house_number",
          },
        ],
        "trading_names": Array [],
      },
    },
    "heading": "Some other company",
    "id": "219999999",
    "meta": Object {
      "Address": "123 ABC Road, Brighton, BN2 9QB",
    },
    "text": null,
  },
]
`;

exports[`useDnbSearch when findCompany() is called without any filters should return all DNB search results 1`] = `
Array [
  Object {
    "canHandleClick": false,
    "data": Object {
      "datahub_company": Object {
        "id": "0fb3379c-341c-4da4-b825-bf8d47b26baa",
        "latest_interaction": Object {
          "created_on": "2019-08-01T18:10:00",
          "date": "2019-08-01",
          "id": "ec4a46ef-6e50-4a5c-bba0-e311f0471312",
          "subject": "Meeting between DIT and Joe Bloggs",
        },
      },
      "dnb_company": Object {
        "address_country": "GB",
        "address_county": "",
        "address_line_1": "123 Fake Street",
        "address_line_2": "",
        "address_postcode": "BN1 4SE",
        "address_town": "Brighton",
        "annual_sales": 1860000000,
        "annual_sales_currency": "USD",
        "domain": "example.co.uk",
        "duns_number": "12345678",
        "employee_number": 2000,
        "global_ultimate_duns_number": "123456789",
        "global_ultimate_primary_name": "Some parent company name",
        "industry_codes": Array [
          Object {
            "usSicV4": "1623",
            "usSicV4Description": "Water/sewer/utility construction",
          },
        ],
        "is_annual_sales_estimated": null,
        "is_employees_number_estimated": true,
        "is_out_of_business": false,
        "legal_status": "corporation",
        "primary_name": "Some company name",
        "registered_address_country": "GB",
        "registered_address_county": "",
        "registered_address_line_1": "",
        "registered_address_line_2": "",
        "registered_address_postcode": "BN1 4SE",
        "registered_address_town": "Brighton",
        "registration_numbers": Array [
          Object {
            "registration_number": "1234567",
            "registration_type": "uk_companies_house_number",
          },
        ],
        "trading_names": Array [
          "Some trading name",
        ],
      },
    },
    "heading": "Some company name",
    "id": "12345678",
    "meta": Object {
      "Address": "123 Fake Street, Brighton, BN1 4SE",
    },
    "text": <React.Fragment>
      This company is already on Data Hub.
       
      <ForwardRef
        href="/companies/0fb3379c-341c-4da4-b825-bf8d47b26baa"
        muted={false}
        noVisitedState={false}
        textColour={false}
      >
        Go to the company page
      </ForwardRef>
       to record activity.
    </React.Fragment>,
  },
  Object {
    "canHandleClick": true,
    "data": Object {
      "datahub_company": null,
      "dnb_company": Object {
        "address_country": "GB",
        "address_county": "",
        "address_line_1": "123 ABC Road",
        "address_line_2": "",
        "address_postcode": "BN2 9QB",
        "address_town": "Brighton",
        "annual_sales": 1999999999,
        "annual_sales_currency": "USD",
        "domain": "example.co.uk",
        "duns_number": "219999999",
        "employee_number": 300,
        "global_ultimate_duns_number": "319999999",
        "global_ultimate_primary_name": "Some other company parent",
        "industry_codes": Array [
          Object {
            "usSicV4": "3799",
            "usSicV4Description": "Mfg transportation equipment",
          },
        ],
        "is_annual_sales_estimated": null,
        "is_employees_number_estimated": true,
        "is_out_of_business": false,
        "legal_status": "corporation",
        "primary_name": "Some other company",
        "registered_address_country": "GB",
        "registered_address_county": "",
        "registered_address_line_1": "",
        "registered_address_line_2": "",
        "registered_address_postcode": "BN2 9QB",
        "registered_address_town": "Brighton",
        "registration_numbers": Array [
          Object {
            "registration_number": "00016033",
            "registration_type": "uk_companies_house_number",
          },
        ],
        "trading_names": Array [],
      },
    },
    "heading": "Some other company",
    "id": "219999999",
    "meta": Object {
      "Address": "123 ABC Road, Brighton, BN2 9QB",
    },
    "text": null,
  },
  Object {
    "canHandleClick": true,
    "data": Object {
      "datahub_company": null,
      "dnb_company": Object {
        "address_country": "GB",
        "address_county": "",
        "address_line_1": "20 New Fake Road",
        "address_line_2": "",
        "address_postcode": "B11 9QC",
        "address_town": "Bristol",
        "annual_sales": 2222222,
        "annual_sales_currency": "USD",
        "domain": "example.co.uk",
        "duns_number": "219999996",
        "employee_number": 100,
        "global_ultimate_duns_number": "319999999",
        "global_ultimate_primary_name": "Dissolvex Global",
        "is_annual_sales_estimated": null,
        "is_employees_number_estimated": true,
        "is_out_of_business": true,
        "legal_status": "corporation",
        "primary_name": "Dissolvex",
        "registered_address_country": "GB",
        "registered_address_county": "",
        "registered_address_line_1": "",
        "registered_address_line_2": "",
        "registered_address_postcode": "B11 9QC",
        "registered_address_town": "Bristol",
        "trading_names": Array [],
      },
    },
    "heading": "Dissolvex",
    "id": "219999996",
    "meta": Object {
      "Address": "20 New Fake Road, Bristol, B11 9QC",
    },
    "text": "This company has stopped trading and is no longer in business.",
  },
]
`;

exports[`useDnbSearch when transformCompanyRecord() is called on a company which is in Data Hub should return transformed company record 1`] = `
Object {
  "canHandleClick": false,
  "data": Object {
    "datahub_company": Object {
      "id": "0fb3379c-341c-4da4-b825-bf8d47b26baa",
      "latest_interaction": Object {
        "created_on": "2019-08-01T18:10:00",
        "date": "2019-08-01",
        "id": "ec4a46ef-6e50-4a5c-bba0-e311f0471312",
        "subject": "Meeting between DIT and Joe Bloggs",
      },
    },
    "dnb_company": Object {
      "address_country": "GB",
      "address_county": "",
      "address_line_1": "123 Fake Street",
      "address_line_2": "",
      "address_postcode": "BN1 4SE",
      "address_town": "Brighton",
      "annual_sales": 1860000000,
      "annual_sales_currency": "USD",
      "domain": "example.co.uk",
      "duns_number": "12345678",
      "employee_number": 2000,
      "global_ultimate_duns_number": "123456789",
      "global_ultimate_primary_name": "Some parent company name",
      "industry_codes": Array [
        Object {
          "usSicV4": "1623",
          "usSicV4Description": "Water/sewer/utility construction",
        },
      ],
      "is_annual_sales_estimated": null,
      "is_employees_number_estimated": true,
      "is_out_of_business": false,
      "legal_status": "corporation",
      "primary_name": "Some company name",
      "registered_address_country": "GB",
      "registered_address_county": "",
      "registered_address_line_1": "",
      "registered_address_line_2": "",
      "registered_address_postcode": "BN1 4SE",
      "registered_address_town": "Brighton",
      "registration_numbers": Array [
        Object {
          "registration_number": "1234567",
          "registration_type": "uk_companies_house_number",
        },
      ],
      "trading_names": Array [
        "Some trading name",
      ],
    },
  },
  "heading": "Some company name",
  "id": "12345678",
  "meta": Object {
    "Address": "123 Fake Street, Brighton, BN1 4SE",
  },
  "text": <React.Fragment>
    This company is already on Data Hub.
     
    <ForwardRef
      href="/companies/0fb3379c-341c-4da4-b825-bf8d47b26baa"
      muted={false}
      noVisitedState={false}
      textColour={false}
    >
      Go to the company page
    </ForwardRef>
     to record activity.
  </React.Fragment>,
}
`;

exports[`useDnbSearch when transformCompanyRecord() is called on a company which is not in Data Hub should return transformed company record 1`] = `
Object {
  "canHandleClick": true,
  "data": Object {
    "datahub_company": null,
    "dnb_company": Object {
      "address_country": "GB",
      "address_county": "",
      "address_line_1": "123 ABC Road",
      "address_line_2": "",
      "address_postcode": "BN2 9QB",
      "address_town": "Brighton",
      "annual_sales": 1999999999,
      "annual_sales_currency": "USD",
      "domain": "example.co.uk",
      "duns_number": "219999999",
      "employee_number": 300,
      "global_ultimate_duns_number": "319999999",
      "global_ultimate_primary_name": "Some other company parent",
      "industry_codes": Array [
        Object {
          "usSicV4": "3799",
          "usSicV4Description": "Mfg transportation equipment",
        },
      ],
      "is_annual_sales_estimated": null,
      "is_employees_number_estimated": true,
      "is_out_of_business": false,
      "legal_status": "corporation",
      "primary_name": "Some other company",
      "registered_address_country": "GB",
      "registered_address_county": "",
      "registered_address_line_1": "",
      "registered_address_line_2": "",
      "registered_address_postcode": "BN2 9QB",
      "registered_address_town": "Brighton",
      "registration_numbers": Array [
        Object {
          "registration_number": "00016033",
          "registration_type": "uk_companies_house_number",
        },
      ],
      "trading_names": Array [],
    },
  },
  "heading": "Some other company",
  "id": "219999999",
  "meta": Object {
    "Address": "123 ABC Road, Brighton, BN2 9QB",
  },
  "text": null,
}
`;

exports[`useDnbSearch when transformCompanyRecord() is called on a dissolved company should return transformed company record 1`] = `
Object {
  "canHandleClick": true,
  "data": Object {
    "datahub_company": null,
    "dnb_company": Object {
      "address_country": "GB",
      "address_county": "",
      "address_line_1": "20 New Fake Road",
      "address_line_2": "",
      "address_postcode": "B11 9QC",
      "address_town": "Bristol",
      "annual_sales": 2222222,
      "annual_sales_currency": "USD",
      "domain": "example.co.uk",
      "duns_number": "219999996",
      "employee_number": 100,
      "global_ultimate_duns_number": "319999999",
      "global_ultimate_primary_name": "Dissolvex Global",
      "is_annual_sales_estimated": null,
      "is_employees_number_estimated": true,
      "is_out_of_business": true,
      "legal_status": "corporation",
      "primary_name": "Dissolvex",
      "registered_address_country": "GB",
      "registered_address_county": "",
      "registered_address_line_1": "",
      "registered_address_line_2": "",
      "registered_address_postcode": "B11 9QC",
      "registered_address_town": "Bristol",
      "trading_names": Array [],
    },
  },
  "heading": "Dissolvex",
  "id": "219999996",
  "meta": Object {
    "Address": "20 New Fake Road, Bristol, B11 9QC",
  },
  "text": "This company has stopped trading and is no longer in business.",
}
`;
